### PR TITLE
feat: list the topic and dates for a dlq-bucket

### PR DESCRIPTION
**Pull request by lavkesh**

### CODE REVIEW

File: entropy/entropy.go
Comment (ishanarya0) [inline entropy/entropy.go:1]: instead of defining these again, let's import it from entropy? https://github.com/goto/entropy/blob/main/modules/firehose/config.go

**Review by ishanarya0 — CHANGES_REQUESTED** [review]

File: entropy/entropy.go
Comment (lavkesh) [reply entropy/entropy.go:1]: yea we tried this, but there was compile-time error. 
```
# oras.land/oras-go/pkg/auth/docker
../../../go/pkg/mod/oras.land/oras-go@v1.2.3/pkg/auth/docker/login.go:86:39: cannot use remote (variable of type *"github.com/docker/docker/registry".Service) as "github.com/docker/docker/registry".Service value in argument to c.loginWithTLS
```
It's ok to create on the client side actually as we need the struct to be unmarshalled. Although this might cause some syncing issues.

**Review by ishanarya0 — CHANGES_REQUESTED** [review]

File: entropy/entropy.go
Comment (ishanarya0) [reply entropy/entropy.go:1]: yeah due to the syncing issue, we moved on using it as a dependency.

also, when and where is the above error happening?

Comment (lavkesh) [reply entropy/entropy.go:1]: it was happening while compiling, seems like this is coming from helm dependency.

Comment (ishanarya0) [reply entropy/entropy.go:1]: okay... though this is not happening on main branch 🤔 
@StewartJingga can you check ^

Comment (StewartJingga) [reply entropy/entropy.go:1]: It may not happening on the entropy branch because of different dependencies that dex and entropy has.
The issue is not actually on the helm but the dependencies helm requires.

File: internal/server/v1/firehose/actions.go [hunk 1]
```diff
@@ -27,7 +26,7 @@ const (
 )
 
 func (api *firehoseAPI) handleReset(w http.ResponseWriter, r *http.Request) {
-	var reqBody entropyKafka.ResetParams
+	var reqBody entropy.ResetParams
```
[the remark this answers is not in the document]
Comment (StewartJingga) [reply internal/server/v1/firehose/actions.go:29]: @ishanarya0 there were dependency issues when importing entropy to dex.

It would be nice if we can use the model directly from entropy, but it seems like it brings more issues because as we are importing entropy to dex, it pulls all the dependencies that entropy uses which most of them are not required by dex.

So for now, dex will define it's own entropy models.

File: .golangci.yml [hunk 2]
```diff
@@ -135,9 +135,10 @@ issues:
...
         - goerr113
+        - nosnakecase
     - path: cli
       linters:
         - forbidigo
         - contextcheck
 severity:
-  default-severity: error
\ No newline at end of file
+  default-severity: error
```
Comment (StewartJingga) [inline .golangci.yml:144]: why is this line marked as changed 🤔

File: .golangci.yml [hunk 3]
```diff
@@ -135,9 +135,10 @@ issues:
         - gocritic
         - dupl
         - goerr113
+        - nosnakecase
```
Comment (StewartJingga) [inline .golangci.yml:138]: we don't need to exclude rule here.

checking from your code, you are building entropy mock manually.
we can build it using library and ideally it should work without excluding rule.

steps:
1. build mock - ([example](https://github.com/goto/dex/blob/main/Makefile#L41))
2. remove the manually built mock in your test
3. use the generated mock - ([example](https://github.com/goto/dex/blob/main/internal/server/v1/alert/subscription_test.go#L38))

File: Makefile [hunk 4]
```diff
@@ -64,3 +64,4 @@ download:
 setup:
 	@go install github.com/vektra/mockery/v2@v2.30.1
 	@go install mvdan.cc/gofumpt@v0.5.0
+	@go install github.com/daixiang0/gci@v0.11.1
```
Comment (StewartJingga) [inline Makefile:67]: thankss for adding this

File: internal/server/v1/dlq/handler_test.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+	args := g.Called(bucketInfo)
+	if args.Get(0) == nil {
+		err, _ := args.Get(1).(error)
+		return nil, err
+	}
+	topicDates, _ := args.Get(0).(map[string]map[string]int64)
+	return topicDates, nil
+}
+
+type mockHTTPWriter struct {
```
Comment (StewartJingga) [inline internal/server/v1/dlq/handler_test.go:78]: do we need to mock http writer? for most cases we should be able to use the one from golang package itself.

example: https://github.com/goto/dex/blob/main/internal/server/v1/alert/routes_test.go#L55

File: internal/server/v1/dlq/handler_test.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+	topicDates["topic-1"]["2023-12-10"] = int64(4321)
+	topicDates["topic-2"]["2023-09-20"] = int64(99)
+	gClient.On("ListTopicDates", gcs.BucketInfo{
+		BucketName: "test-bucket",
+		Prefix:     "test-prefix",
+		Delim:      "",
+	}).Return(topicDates, nil)
+	handler.listFirehoseDLQ(httpWriter, httpRequest)
+	expectedMap := make(map[string]map[string]map[string]int64)
+	_ = json.Unmarshal([]byte(httpWriter.messages[0]), &expectedMap)
```
Comment (StewartJingga) [inline internal/server/v1/dlq/handler_test.go:154]: let's always try to handle error. in this case, we can easily do
```suggestion
	err = json.Unmarshal([]byte(httpWriter.messages[0]), &expectedMap)
	require.NoError(t, err)
```

this one line can save us a few mins debugging

File: internal/server/v1/dlq/handler_test.go [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+	assert.Equal(t, topicDates, expectedMap["dlq_list"])
```
Comment (StewartJingga) [inline internal/server/v1/dlq/handler_test.go:155]: this is optional, but another way to assert the response, you can do it like [here](https://github.com/goto/dex/blob/main/internal/server/v1/alert/routes_test.go#L65-L70).

File: internal/server/v1/dlq/handler_test.go [hunk 8]
```diff
@@ -0,0 +1,231 @@
+package dlq
```
Comment (StewartJingga) [inline internal/server/v1/dlq/handler_test.go:1]: let's use `dlq_test` for test package, reason is, we don't want the `dlq` package to share any private variables with the test, in this case, the `dlq` package will have access to `mock*` structs that are here.


Secondly, if your reason for this is because handler methods are private then a few approaches can be: 1. make handler's methods public or 2. test the router, not the handler.

Personally I would just test whatever the package is exposing, and in this current setup, Router is the output of this package, handler is the implementation details, hence I would test the router instead. example: https://github.com/goto/dex/blob/main/internal/server/v1/alert/routes_test.go#L37

**Review by StewartJingga — CHANGES_REQUESTED** [review]

File: internal/server/v1/dlq/handler_test.go (hunk 8 above)
Comment (lavkesh) [reply internal/server/v1/dlq/handler_test.go:1]: hm. in golang we write tests in the same package. let's discuss this?

Comment (StewartJingga) [reply internal/server/v1/dlq/handler_test.go:1]: if you are doing whitebox testing then yes.
But most of the times, we only want to test exported API from the package we are testing. (e.g. blackbox testing)

Using different package will make sure:
1. test can only access exported APIs which simulates how the package is being used by other packages using it
2. no variables conflict - you can create any vars or const to help refactor your test without having to interfere with application package (e.g. `dlq`)

So basically there is no hard rule per se, you can still use same package, but I don't see any benefits doing that unless I want to have extensive tests on private APIs. (e.g. testing complex logic but the API is not exposed)

File: .golangci.yml (hunk 3 above)
Comment (lavkesh) [reply .golangci.yml:138]: this is nice!

File: internal/server/v1/dlq/handler_test.go (hunk 5 above)
Comment (lavkesh) [reply internal/server/v1/dlq/handler_test.go:78]: this is fine, will rename to testHttpRequest. we are no running a mock server.

File: internal/server/v1/dlq/handler_test.go (hunk 7 above)
Comment (lavkesh) [reply internal/server/v1/dlq/handler_test.go:100]: maybe when we write route testcases we can use this.

File: Makefile [hunk 9]
```diff
@@ -41,6 +41,10 @@ generate-mocks:
 	@mockery --srcpkg=buf.build/gen/go/gotocompany/proton/grpc/go/gotocompany/siren/v1beta1/sirenv1beta1grpc --name=SirenServiceClient
 	@mockery --srcpkg=buf.build/gen/go/gotocompany/proton/grpc/go/gotocompany/shield/v1beta1/shieldv1beta1grpc --name=ShieldServiceClient
 	@mockery --srcpkg=buf.build/gen/go/gotocompany/proton/grpc/go/gotocompany/optimus/core/v1beta1/corev1beta1grpc --name=JobSpecificationServiceClient
+	@mockery --srcpkg=buf.build/gen/go/gotocompany/proton/grpc/go/gotocompany/entropy/v1beta1/entropyv1beta1grpc --name=ResourceServiceClient
+	@mockery --srcpkg=./internal/server/gcs --name=BlobStorageClient
```
Comment (StewartJingga) [inline Makefile:45]: for locally defined interface, there is a better way to do it.
I put Siren the rest rpc clients here because they are remote files.

But let's do this approach for now, we can refactor later if it get bigger.

**Review by StewartJingga — APPROVED** [review]
LGTM

File: internal/server/gcs/models.go [hunk 10]
```diff
@@ -8,7 +8,13 @@ import (
 
 // BlobStorageClient This is used in service
 type BlobStorageClient interface {
-	ListTopicDates(bucketInfo BucketInfo) (map[string]map[string]int64, error)
+	ListTopicDates(bucketInfo BucketInfo) ([]TopicMetaData, error)
+}
+
+type TopicMetaData struct {
```
Comment (StewartJingga) [inline internal/server/gcs/models.go:14]: you can use model from here
https://github.com/goto/dex/blob/main/generated/models/dlq.go#L20
we define our domain model via swagger and generated there.

Defining the model in gcs package and use it as response might not seem scalable.
Response might be enriched in the future, but it may not come from gcs.

**Review by ishanarya0 — APPROVED** [review]

**Review by StewartJingga — APPROVED** [review]
